### PR TITLE
[Pending mbed-os 5.6] Print peripheral on ACL violation from debug blinky example

### PR DESCRIPTION
**Pull request by theamirocohen**
Print the name of a peripheral when faulting on an address belonging
to a perhipheral.
The memory maps of K64F and STM32F4 were added.

### CODE REVIEW

**Comment by theamirocohen** [comment]
@danny4478 
@romkuz01 
please review

File: source/mem_map/mem_map.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __MEM_MAP_H__
+#define __MEM_MAP_H__
+
+typedef struct mem_map {
+    const char *name;
```
Comment (danny4478) [inline source/mem_map/mem_map.h]: Space after *\

File: source/mem_map/mem_map.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    uint32_t base;
+    uint32_t end;
+} MemMap;
+
+const MemMap* memory_map_name(uint32_t addr);
```
Comment (danny4478) [inline source/mem_map/mem_map.h]: Space after *

File: source/mem_map/mem_map.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#include "plat_mem_map.h"
+
+const MemMap* memory_map_name(uint32_t addr)
+{
+    int i;
+    const MemMap *map;
+
+    /* find system memory region */
+    map = g_mem_map;
+    for(i = 0; i < UVISOR_ARRAY_COUNT(g_mem_map); i++)
```
Comment (danny4478) [inline source/mem_map/mem_map.cpp]: I would always use " { ... }", even wen not needed by compiler.
It is safer.

File: source/main.cpp [hunk 4]
```diff
@@ -25,6 +26,24 @@ MAIN_ACL(g_main_acl);
 UVISOR_SET_MODE_ACL(UVISOR_ENABLED, g_main_acl);
 UVISOR_SET_PAGE_HEAP(8 * 1024, 5);
 
+static void example_halt_error(THaltError reason, const THaltInfo *halt_info);
+
+UVISOR_PUBLIC_BOX_DEBUG_DRIVER(example_halt_error);
+
+static void example_halt_error(THaltError reason, const THaltInfo *halt_info) {
+
+	const MemMap *map;
```
Comment (danny4478) [inline source/main.cpp]: Space after  *

File: source/main.cpp [hunk 5]
```diff
@@ -25,6 +26,24 @@ MAIN_ACL(g_main_acl);
...
+	map = memory_map_name(halt_info->bfar);
```
Comment (danny4478) [inline source/main.cpp]: Are we sure here that bfar is valid?

File: source/mem_map/mem_map.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        if((addr >= map->base) && (addr <= map->end))
+            return map;
```
Comment (alzix) [inline source/mem_map/mem_map.cpp]: same goes for `if` statements - braces ARE mandatory for all block  statements - also for single line cases

**Comment by Patater** [comment]
"Print peripheral on ACL violation from debug fault example"

Shouldn't this PR be against the [debug fault example](https://github.com/ARMmbed/mbed-os-example-uvisor-debug-fault)?

**Comment by theamirocohen** [comment]
Hi,

@alzix instructed me to use the basic uVisor example from uvisor repo

**Comment by Patater** [comment]
Please update the commit message to reflect this, then.

Thanks

**Comment by alzix** [comment]
Hi @Patater ,
our debug-fault example faults on arbitrary address - thus this addition would not be helpful at all.
On the other hand mbed-os-example-uvisor is mentioned in our porting guide - thus porting engineers most likely will start with it and at first stages run in to some faults.
We are open for suggestions.

**Comment by Patater** [comment]
@alzix Makes sense to me. Not sure how valuable the separate debug fault example is with this change, though.

**Comment by mikisch81** [comment]
@Patater I think the debug fault example is still useful as it's an example which explicitly shows a fault scenario.
The *mbed-os-example-uvisor* normally should not fault, BUT if it does faults (usually because of missing ACL entry) we need to show what is the expected output.

**Comment by Patater** [comment]
@mikisch81 @alzix 
Our old uVisor example, where pressing the button made the system try to access something it shouldn't and fault, would be a good place for a debug box example. But, the memory map isn't really needed there, as the example touches another box's SRAM and not a peripheral.

It'd probably be best to modify our debug fault example to poke some unowned peripheral instead of 0xFFFFFFFF, that way we could use a memory map like this. The IRQ example could stay simple as well without the memory map, although I agree with the porting benefits of having the memory map here too.

**Comment by theamirocohen** [comment]
pending for mbed-os : mbed-os-5.6.0-rc1
depends on THaltInfo struct

**Comment by alzix** [comment]
retest uvisor

**Comment by alzix** [comment]
@theamirocohen , please update mbed-os hash to point after https://github.com/ARMmbed/mbed-os/pull/5275 merge

**Comment by theamirocohen** [comment]
@alzix, I've updated mbed-os hash to point after ARMmbed/mbed-os#5275 merge

**Pull request merged**